File: dumpir.py
#!/usr/bin/env python

import sys

from parser import parse, optimize
from parser import OUTPUT, INPUT, LOOP, ENDLOOP, MOVE
from parser import ADD, SET, MULCOPY, SCAN, LOADOUT, LOADOUTSET
from parser import IF, ENDIF


def dumpir(code):
    tokens = parse(code)    
    tokens = optimize(tokens)
    for token, value in tokens:
        if token == INPUT:
            print('input')
        elif token == LOOP:
            print('loop')
        elif token == ENDLOOP:
            print('endloop')
        elif token == IF:
            print('if(offset=%d)' % value)
        elif token == ENDIF:
            print('endif')
        elif token == MOVE:
            print('move(offset=%d)' % value)
        elif token == ADD:
            print('add(offset=%d, value=%d)' % value)
        elif token == SET:
            print('set(offset=%d, value=%d)' % value)
        elif token == MULCOPY:
            print('mulcopy(src=%d, dest=%d, mul=%d)' % value)
        elif token == SCAN:
            print('scan(offset=%d)' % value)
        elif token == LOADOUT:
            print('loadout(offset=%d, add=%d)' % value)
        elif token == LOADOUTSET:
            print('loadoutset(value=%d)' % value)
        elif token == OUTPUT:
            print('output')
        else:
            print('TOKEN', i, 'NOT HANDLED')

if __name__ == '__main__':
    with open(sys.argv[1]) as bffile:
        dumpir(bffile.read())

File: src/bin/dumpir.rs
use std::env;
use std::io::prelude::*;
use std::fs::File;

extern crate isbfc;
use isbfc::Token;
use isbfc::parse;
use isbfc::optimize;

fn main() {
    let path = env::args().nth(1).unwrap();
    let mut file = File::open(path).unwrap();
    let mut code = String::new();
    file.read_to_string(&mut code).unwrap();
    
    let tokens = parse(&code);
    let tokens = optimize(tokens);

    for token in tokens.iter() {
        match *token {
            Token::Output =>
                println!("output"),
            Token::Input =>
                println!("input"),
            Token::Loop =>
                println!("loop"),
            Token::EndLoop =>
                println!("endloop"),
            Token::Move(offset) =>
                println!("move(offset={})", offset),
            Token::Add(offset, value) =>
                println!("add(offset={}, value={})", offset, value),
            Token::Set(offset, value) =>
                println!("set(offset={}, value={})", offset, value),
            Token::MulCopy(src, dest, mul) =>
                println!("mulcopy(src={}, dest={}, mul={})", src, dest, mul),
            Token::Scan(offset) =>
                println!("scan(offset={})", offset),
            Token::LoadOut(offset, add) =>
                println!("loadout(offset={}, add={})", offset, add),
            Token::LoadOutSet(value) =>
                println!("loadoutset(value={})", value),
            Token::If(offset) =>
                println!("if(offset={})", offset),
            Token::EndIf =>
                println!("endif"),
        }
    }
}
